pkg/watch/versioned/decoder.go

```diff
@@ -14,56 +14,54 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package json
+package versioned
 
 import (
-	"encoding/json"
 	"fmt"
-	"io"
 
 	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer/streaming"
 	"k8s.io/kubernetes/pkg/watch"
 )
 
 // Decoder implements the watch.Decoder interface for io.ReadClosers that
-// have contents which consist of a series of watchEvent objects encoded via JSON.
-// It will decode any object registered in the supplied codec.
+// have contents which consist of a series of watchEvent objects encoded
+// with the given streaming decoder. The internal objects will be then
+// decoded by the embedded decoder.
 type Decoder struct {
-	r       io.ReadCloser
-	decoder *json.Decoder
-	codec   runtime.Codec
+	decoder         streaming.Decoder
+	embeddedDecoder runtime.Decoder
 }
 
 // NewDecoder creates an Decoder for the given writer and codec.
-func NewDecoder(r io.ReadCloser, codec runtime.Codec) *Decoder {
+func NewDecoder(decoder streaming.Decoder, embeddedDecoder runtime.Decoder) *Decoder {
 	return &Decoder{
-		r:       r,
-		decoder: json.NewDecoder(r),
-		codec:   codec,
+		decoder:         decoder,
+		embeddedDecoder: embeddedDecoder,
 	}
 }
 
-// Decode blocks until it can return the next object in the writer. Returns an error
-// if the writer is closed or an object can't be decoded.
+// Decode blocks until it can return the next object in the reader. Returns an error
+// if the reader is closed or an object can't be decoded.
 func (d *Decoder) Decode() (watch.EventType, runtime.Object, error) {
-	var got WatchEvent
-	if err := d.decoder.Decode(&got); err != nil {
+	var got Event
+	if _, _, err := d.decoder.Decode(nil, &got); err != nil {
```

Comment:
You need to check and use the returned object - the contract of decode is
that the returned object might be different from the into object.

On Fri, Apr 29, 2016 at 3:28 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/watch/versioned/decoder.go
> <https://github.com/kubernetes/kubernetes/pull/24789#discussion_r61542187>
> :
>
> >  func (d *Decoder) Decode() (watch.EventType, runtime.Object, error) {
> > -	var got WatchEvent
> > -	if err := d.decoder.Decode(&got); err != nil {
> > +	var got Event
> > +	if _, _, err := d.decoder.Decode(nil, &got); err != nil {
>
> However, I'm passing the "into" (got) as an argument and it is supposed to
> check if type is ok.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24789/files/46273cb4bc5066abfd55fcd46fb04e61e8b5b38a#r61542187>
>